src/bootstrap/compile.rs

```diff
@@ -854,13 +854,25 @@ pub fn run_cargo(
 
             let filename = Path::new(&*filename);
 
-            // If this was an output file in the "host dir" we don't actually
-            // worry about it, it's not relevant for us
+            // If this was an output file in the "host dir", we want
+            // to ignore it if it's a build script.
             if filename.starts_with(&host_root_dir) {
-                // Unless it's a proc macro used in the compiler
                 if crate_types.iter().any(|t| t == "proc-macro") {
                     deps.push((filename.to_path_buf(), true));
                 }
+
+                let is_build_script = kind == &["custom_build"]
+                    && crate_types == &["bin"]
+                    && name == "build-script-build";
+                // We don't care about build scripts, but we *do* care about proc-macro
+                // dependencies - the compiler needs their metadata when loading proc-macro
+                // crates. Anything that's not a build script should be a proc-macro dependency.
+                //
+                // FIXME: Have Cargo explicitly indicate build-script vs proc-macro dependencies,
+                // instead of relying on this check.
+                if !is_build_script {
+                    deps.push((filename.to_path_buf(), false));
+                }
```

Comment:
Why does this need a different check from the above `crate_types.iter().any(|t| t == "proc-macro")`?

Generally it also feels wrong or at least odd that we're putting proc macros seemingly twice, once into the host and once into the target directory. That feels reminiscent of a flag @Zoxc, IIRC, added (-Zdual-macros, or so, I don't recall exactly).

Reply:
The `crate_types.iter().any(|t| t == "proc-macro")` check is looking for proc macro crates. This check is looking for *dependencies* of proc-macros, which may not be proc-macros themselves.

For example, suppose we have the dependency graph `normal_crate_one` -> `proc_macro_crate` -> `normal_crate_two`. The crate `normal_crate_two` is not a proc macro crate - but since it's a dependency of `proc_macro_crate`, it will end up in the host dir.